docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+
+1.  A cluster has only a single provisioner at a time, dictated by the cloud
+    provider
+2.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+3.  If no volume is found for the claim, and the claim has the annotation
+    `volume.alpha.kubernetes.io/storage-class`, the provisioner is invoked
+    to provision a volume to fill that claim
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin` which has a method, `NewProvisioner()`,
+    that returns a new `Provisioner` instance
+5.  The controller calls the provisioner's `Provision()` method; `Provision`
+    is responsible for provisioning the volume in the cloud provider and
+    returns an `api.PersistentVolume` and an error
+6.  If `Provision` returns an error, the controller stops trying to provision
+7.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, already bound to the claim
+  1.  If the create operation fails, it is retried
+  2.  If the create operation never succeeds, the controller attempts to delete
+      the provisioned volume and creates an event on the claim
+
+### When should we provision volumes?
+
+It is sensible to provision volumes only when an existing volume cannot
+satisfy a claim.  This ensures that existing, already allocated resources are
+used before additional resources are allocated.  In this regard, the existing
+logic need not change much when storage classes are added.  The primary
+difference is that in the existing design, there is no way to parameterize
+provisioners or have more than one provisioner per cluster.  Therefore, once
+we know that we have to provision a volume to satisfy a claim, we need to
+determine which storage class to use for the claim.
+
+### How should a claim be matched to a storage class?
+
+We've discussed so far that volume claims and storage classes will be matched
+using the claim's label selector.  With that said, there are a number of small
+details to think about:
+
+1.  What should happen if there are multiple storage classes that match a
+    claim's selector?
+2.  Should we provide a way to specify a default storage class for claims
+    without a selector
+3.  Should we provide a way to specify a default storage class for claims
+    whose selector does not match any storage classes?
+
+For now, we will think of the problem of matching a claim to a storage class
+as a black box that solves the problem of: _given a claim and a list of
+storage classes that match that claim's selector, find a storage class to
+provision a volume for the claim with_.  Initially, we will implement the
+following behavior for this black box:
+
+1.  If multiple storage classes match the claim's selector, choose one at
```

Comment:
If we go the route of single "storage-class" label instead of multiple labels, we won't have this confusion and would be more deterministic.